src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -70,31 +90,30 @@ public async ValueTask DisposeAsync()
             }
 
             _disposed = true;
-            _queuedCommandsWriter.Complete();
-            await _pipeWriter.CompleteAsync().ConfigureAwait(false);
+
+            if (_pipeWriter != null)
+                await _pipeWriter.CompleteAsync().ConfigureAwait(false);
+            if (_pipeReader != null)
+                await _pipeReader.CompleteAsync().ConfigureAwait(false);
+
+            await _readerLoopTask.ConfigureAwait(false);
         }
         finally
         {
-            _semLock.Release();
+            while (!_channelLock.Reader.TryRead(out _))
```

Comment:
so would you say using ReadAsync() would be better?

Reply:
🤔 

1. Need to think more about the `.Dispose()` but probably yes.
2. I think in general, `ReadAsync` should be used (unless a strong argument exists for thrashing on non-critical ops)
3. I'll make a sample helper for the pattern I'm thinking of (or pull branch and provide commit link or draft pr against your branch if that's ok and I get the opportunity to)

Where using a `BoundedChannel` vs a `SemaphoreSlim` can be challenging, is in deciding, for lack of better words, 'the right pivot' on how to use the channel.

Depending on the need, and how the underlying synchronization abstraction is implemented, can have a sizeable impact on performance.

Unless there is a compelling reason to do otherwise I think I'm settled on `.ReadAsync` being a better idea for what we are doing here.

If we need more, we can go deeper; `BoundedChannel` is lower alloc than `SemaphoreSlim` for _many_ cases, however there may be future opportunities for a better pooled abstraction[0]


[0] - In general, if a channel has a waiting value on read, it's no alloc. if it's the FIRST waiter with nondefault cancellation token, it gets a pooled instance. if there's a token or the pooled waiter is used, alloc. A _little_ better than semaphoreslim and has in some ways more reliable semantics, but it can be tricky at first.